core/base.py

```diff
@@ -0,0 +1,323 @@
+
+        for property_key in self.required_properties:
+            ec2_utils.validate_node_property(property_key, ctx.node.properties)
+
+        if self.external_resource and not resource:
+            raise NonRecoverableError(
+                'External resource, but the supplied {0} does not exist in the account.'.format(self.aws_resource_type))
+
+        if not self.external_resource and resource:
+            raise NonRecoverableError(
+                'Not external resource, but the supplied {0} exists in the account.'.format(self.aws_resource_type))
+
+    def create(self):
+        return False
+
+    def created(self):
+
+        if self.create_external_resource():
+            return self.post_create()
+        elif self.create():
+            return self.post_create()
+
+        raise NonRecoverableError('Unable to create this resource.')
+
+    def create_external_resource(self):
+
+        if not self.external_resource:
+            return False
+
+        resources = self.get_all_matching([self.resource_id])
+
+        if not resources:
+            self.raise_cannot_use_external_resource(self.resource_id)
+
+        ec2_utils.set_external_resource_id(self.resource_id, ctx.instance)
+
+        return True
+
+    def delete(self):
+        return False
+
+    def deleted(self):
+
+        if not self.get_resource():
+            raise NonRecoverableError('resource is not in the account: {0}.'.format(self.resource_id))
+
+        if self.delete_external_resource():
+            return self.post_delete()
+        elif self.delete():
+            return self.post_delete()
+
+        raise NonRecoverableError('Unable to delete this resource.')
+
+    def delete_external_resource(self):
+
+        if not self.external_resource:
+            return False
+
+        ctx.logger.info('External resource. Not deleting resource {0}.'.format(self.resource_id))
+
+        return True
+
+    def get_all_matching(self, list_of_ids=None):
+
+        args = {self.get_all_argument: list_of_ids}
+
+        try:
+            resources = self.get_all_function(**args)
+        except exception.EC2ResponseError as e:
+            if self.InvalidNotFoundError in str(e):
+                resources = self.get_all_function()
+                ec2_utils.log_available_resources(resources)
+                return None
+            raise NonRecoverableError('{0}'.format(str(e)))
+        except exception.BotoServerError as e:
+            raise NonRecoverableError('{0}'.format(str(e)))
+
+        return resources
+
+    def get_resource(self):
+
+        list_of_matching_resources = self.get_all_matching([self.resource_id])
+
+        if list_of_matching_resources:
+            for resource in list_of_matching_resources:
+                if resource.id == self.resource_id:
+                    return resource
+
+        return None
+
+    def post_create(self):
+        ec2_utils.set_external_resource_id(self.resource_id, ctx.instance)
```

Comment:
talk to idan about all this External not external external aws cloudify stuff because it's so confusing.